src/jitlayers.cpp

```diff
@@ -349,15 +350,21 @@ jl_code_instance_t *jl_generate_fptr_impl(jl_method_instance_t *mi JL_PROPAGATES
             if (src->inferred && !codeinst->inferred)
                 codeinst->inferred = jl_nothing;
         }
+        // heuristic to identify whether an invalidated method is being recompiled
+        is_recompile = codeinst->min_world == world;
```

Comment:
Why does this work? If it's a fresh compile, isn't `min_world` also set to `world`?

I would do this up higher: after https://github.com/JuliaLang/julia/blob/1600cb97d2e829f87b80295c0b01e6b1cf516906/src/jitlayers.cpp#L326 I would add an `else is_recompile = mi->cache != NULL`. What that does is test whether `mi` has one or more `CodeInstance`s but none of them apply (see the implementation of `jl_rettype_inferred`.

Reply:
I think I was confused by the way mi's are invalidated. Thanks for the pointer! Seems to work, and a clean `using CSV` reports no recompilation now.